### DISCUSSION TITLE
Timeline Trace in Training

### DISCUSSION THREAD

**Discussion by npiroozan** [deepmodeling/deepmd-kit#1888] [Q&A]
Hello,

I hope you are doing well.  Attached please find the timeline.json script for a training run on a single node with 16 workers and a local batch size of 32.  TF_Intra_Op is set to 12 and TF_Inter_Op is set to 4, with OpenMP=3.  I expected to see 4 compute threads, reflecting TF_Inter_Op setting.

The strange element of this run is on threads 4-9 with so much time spent on HorovodAllReduce.  Is this behavior that you have noticed as well?

In addition, within the Json script there are functions for "enable_profiler" and "profiling" which give differing outputs with respect to the tensorboard trace.  May I also ask what the difference between these two functionalities are?

Thank you very much for your time!

Warm Regards
<img width="959" alt="Timeline_Trace" src="https://user-images.githubusercontent.com/87953329/188034417-de389f23-1681-412f-98c8-d0b2186ffa9d.png">

**Comment by wanghan-iapcm** [comment]
@shishaochen Could you please take a look? Thanks!

**Reply by npiroozan** [reply]
Hello,

I am following up on this point from several weeks ago.  Hope everyone is doing well!

Thank you.

**Comment by njzjz** [comment]
`enable_profiler` works with [TensorFlow Profiler](https://www.tensorflow.org/guide/profiler) which is a new feature in TensorFlow v2.5.

For horovod issue, the behavior should come from horovod itself but not from deepmd-kit.

**Reply by njzjz** [reply]
So, did you use horovod or mpi? It is strange if it was not actually used but consumed time..

**Reply by npiroozan** [reply]
I used mpi to run DeePMD for training.  I'm optimizing on a single node and 16 workers.

Apologies for the late reply, I missed your question.

**Reply by npiroozan** [reply]
Apologies for the follow-up.  Have you seen this behavior in any of your runs (regarding Horovod)?

Thank you!

**Reply by njzjz** [reply]
Do you mean `mpirun -n 16`? How many CPU cores do you have?

**Reply by njzjz** [reply]
This does show that the communications between horovod workers are quite expensive.

**Reply by npiroozan** [reply]
Yes, I was using mpirun -n 16.  On this particular run, I used 16 mpi processes and a local batch size of 32.  Indeed, the expense of the horovod worker streams are my primary concern.

**Reply by njzjz** [reply]
Do you have $16 \times 4 \times 3$ cores? If not, threads will compete for resources.

**Reply by npiroozan** [reply]
In sorry I'm a bit confused by what you mean here.  To answer your question, I don't believe so.

**Reply by npiroozan** [reply]
CPU I'm using has 72 cores

**Reply by njzjz** [reply]
You have 16 workers (via mpi), 4 streams/worker (via TF_Inter_Op), 3 threads/stream (via openmp), so the total number of threads is $16 \times 4 \times 3$.

**Reply by njzjz** [reply]
I can see 16 `HorovodAllReduce` ops in the timeline and it should be the expected. But I don't know why horovod uses 9 streams.

**Reply by npiroozan** [reply]
Understood.  Yes that's something I'm trying to determine also. It does not seem to be impacted with newer version of Horovod.  The image above is derived from horovod 0.24.0.
